### ISSUE TITLE
Hiesenbug: Runtime server initialisation

### ISSUE THREAD

**Issue by ActuallyHappening** [lifescapegame/bevy_replicon#107]
Thanks so much for this awesome crate!

I have adapted the `simple_box` example to use runtime server / client initialisation and am encountering a really strange bug, where consistently one 'part' of my program seems to run at 'the wrong time' so that the server side of `bevy_replicon` doesn't actually replicate any entities to clients, and the other 'part' of my program seems to run 'at the correct time' so that the `bevy_replicon` server works just fine. I will continue to be more specific as I strip away more layers.

I have been frustrated with this for a few days now, and was wondering what errors I might not know about when I initialise my `NetcodeServerTransport` and `RenetServer` resources in the `Update` or `OnEnter(T)` schedule versus the `Startup` schedule. The examples use the `Startup` schedule because it decides if its is a server / client / solo by reading CLI args, I envisage my game to be able to freely switch between server client and solo modes at runtime.

How can I debug `bevy_replicon`? Is it designed for this use case, dynamically switching and connecting to different servers? (Also, I am getting a segfault when running on `nightly` compiler immediately after running so I'm using the stable toolchain)

I just upgraded to `bevy_replicon = "0.16"`.

**Comment by Shatur** [comment]
Hi!

> How can I debug bevy_replicon

We currently have `ClientDiagnosticPlugin` that shows statistic about replicated entities. But nothing else. I think we should add some `trace!` when we send data to clients and receive it. This should help you a bit.

> Is it designed for this use case, dynamically switching and connecting to different servers?

Yes, it does. There shouldn't be any problem. Could you try to provide a minimal reproducible example?

> Also, I am getting a segfault when running on nightly compiler immediately after running so I'm using the stable toolchain

We test it on CI with a stable toolchain too.

**Comment by Shatur** [comment]
@ActuallyHappening try #108 and enable tracing:
```rust
.add_plugins(DefaultPlugins.set(LogPlugin {
    filter: "bevy_replicon=trace".into(),
    level: bevy::log::Level::DEBUG,
})
```
This may help you to debug the problem.

**Comment by ActuallyHappening** [comment]
Thanks so much!
I have 'isolated' the bug into one line which, when uncommented, results in the bug but when commented out everything seems to work. It is `app.replicate::<MyComplexType>()`. This bug is quite elusive, for example I copied most of the code from lib.rs into main.rs and it disappeared!
I'll focus all my attention and clone the project into a seperate package, and strip away layers until it is a minimal reproducible bug. I'm not one to leave a loose bug lying around! (PS: Repo here, although its still messy: https://github.com/ActuallyHappening/creativity-game)

Thanks for the quick reply and PR, I'll patch into the 'trace-replication' branch and send the results.

**Comment by Shatur** [comment]
>  I have 'isolated' the bug into one line which, when uncommented, results in the bug but when commented out everything seems to work. It is app.replicate::<MyComplexType>(). This bug is quite elusive, for example I copied most of the code from lib.rs into main.rs and it disappeared!

That's odd. Maybe you trigger undefined behavior somehow by unsafe code in your repo?

**Comment by ActuallyHappening** [comment]
I have no unsafe code

**Comment by ActuallyHappening** [comment]
OK, this is by no means 'minimal' but is consistently reproducible on my machine.
I will continue to shave off unnecessary files and code until I can isolate the bug. Note the line I was talking about before is in bugged_replicon/core.rs line ~52 with a todo comment above it.

```bash
git clone https://github.com/ActuallyHappening/creativity-game.git
cd creativity_game/bugged_replicon
git checkout bugged # switches branches

# then run two copies of the program whichever way you like
cargo r&
cargo r
```

in one copy of program, press the BLACK 'host' button (and then the next button after that to start hosting) and observe the log in the console `2023-11-01T06:45:14.301193Z  INFO creativity_game_bugged: Setting up server resources: server: [ChannelConfig { channel_id: 0, max_memory_usage_bytes: 5242880, send_type: Unreliable }]; client: [ChannelConfig { channel_id: 0, max_memory_usage_bytes: 5242880, send_type: Unreliable }]`
<img width="347" alt="image" src="https://github.com/lifescapegame/bevy_replicon/assets/105958073/a8eaee1c-95d3-49be-8c7b-5846c654a081">
<img width="326" alt="image" src="https://github.com/lifescapegame/bevy_replicon/assets/105958073/03fa23c2-2819-45bc-ab06-654c57b8e1ed">
<img width="1439" alt="image" src="https://github.com/lifescapegame/bevy_replicon/assets/105958073/0a5030f2-e054-434f-bfd8-dcd5703d76a3">

(Images showing how to setup bugged server)

This server is bugged. If you connect a client copy to it, it seems no entities through `bevy_replicon` are actually replicated. On the server, press the GREEN 'Add' button to add some dummy components/entities to the world.

Then open a client copy with `cargo r` again (in a seperate terminal)

<img width="297" alt="image" src="https://github.com/lifescapegame/bevy_replicon/assets/105958073/1201aa18-c694-4b55-8286-e7d0109deea8">
<img width="1440" alt="image" src="https://github.com/lifescapegame/bevy_replicon/assets/105958073/ae9311d0-a4c6-4c79-9cbe-3b9f3abdf413">

(Images showing how to setup client server, which doesn't seem to be the problem although I am not sure)

Observer the client console (right) and server console (left), with the server registering that a client has connected.
![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/0c05a012-8cf2-45fd-bb76-d20d8fa88dc4)
The number of dummy entities on the client is 0, because replication doesn't seem to work with the BLACK button hosting.

Now, close both windows and start a fresh server by pressing the GREEN 'Host' button, and starting a client as before. Pressing the green 'add' button now syncs into the console.

**Comment by ActuallyHappening** [comment]
Upon applying #108, I get these logs in the bad server but the 'sending replication message' doesn't keep sending to the clients in the good server.
![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/dde689ca-d432-4c52-9d03-0a025de7c37b)

**Comment by Shatur** [comment]
I will be able to take a deep look later, but if you have server sending messages in the log and don't see client receiving any messages, then something wrong with the client. I afraid that the problem could be on Renet side. How big your complex type? Maybe it exceeds the channel memory limit?

**Comment by ActuallyHappening** [comment]
That is definitely possible, I haven't considered it. It contains rather large vectors. If this is the problem, can I increase the channel memory limit? Can I handle the errors somehow? The stack size is only 24 bytes but serialised it would be much larger, maybe a few kilobytes per component.
![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/6d26ebc1-2abf-4b89-bf37-08e44cccce90)

Although, I don't think this is the *whole* root problem because the behaviour changes depending on the method of initialising the server.

**Comment by Shatur** [comment]
> If this is the problem, can I increase the channel memory limit?

Yes, I added the API for it in the latest version.

> Although, I don't think this is the whole root problem because the behaviour changes depending on the method of initialising the server.

I agree, this is suspicious, but maybe in the later scenario client receives the message twice. It's sad that Renet doesn't warn about it. But try to increase the memory limit or reduce the sending data just to test if it's the case.

**Comment by ActuallyHappening** [comment]
Yes, if I don't add any of the `SpawnChildStructure` components into the server world, the bug does not surface, but if I do then the 'bad' server doesn't replicate to the client. For reference, here are the console logs I get when running client with bevy_replicon=trace and renet=trace for the good and the bad server respectively when spawning in `SpawnChildStructure` components/entities: (left is server, right is client, using commit 5d7c6b2c2f3497efc034ec75db35b8be1f2fb2d3 https://github.com/ActuallyHappening/creativity-game.git branch 'bugged' in folder bugged_replicon/Cargo.toml)

![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/e5db6e1d-6b67-4421-8cda-95724403308c)
(Image above shows the 'good' server on the left with a few added dummy components, and the client on the right correctly receiving the dummy components as shows by the INFO log 'Number of dummy components: 17'

![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/d853f46b-c6d2-44f7-9b0c-a9b721c70c2f)

The image above shows the 'bad' server on the left with a few added dummy components, and the client on the right *not* receiving any of the dummy components, presumably because of the spawned `SpawnChildStructure` components which are rather large.

**Comment by Shatur** [comment]
So the memory is the problem. Try increasing memory limit for the replication channel.
And I would suggest to report the issue to Renet to log memory limit errors.

**Comment by ActuallyHappening** [comment]
```rs
app
			.add_plugins(ReplicationPlugins)
			.add_systems(Startup, |mut network_config: ResMut<NetworkChannels>| {
				network_config.set_default_max_bytes(500 * 1024 * 1024);
			})
```

Is this how you do that? It didn't help my specific case.
Yes, I will flag issue with renet

**Comment by Shatur** [comment]
> Is this how you do that? It didn't help my specific case.

Yes, I would expect this to work. If it doesn't help, then I'm not sure. Could you try to serialize your huge struct and send it over Renet without replicon? Just via Renet. Just modify one of the examples.

**Comment by ActuallyHappening** [comment]
It seems to send just fine using `bincode`, will post screenshots in a sec

**Comment by Shatur** [comment]
Could you check how much do you send in replicon? You could see it in statistic or patch the crate and add extra `trace`.

**Comment by ActuallyHappening** [comment]
I (though I) have already patched 'bevy_replicon' and those screenshots of my logs were all I got. Link to example/docs for what you mean?

**Comment by Shatur** [comment]
> Logging the number of bytes I send I get between 1 and 50 thousand

Try to send similar amount of bytes in the Renet's echo example. I think you should be able to reproduce the problem.

> Obviously it is unreliable, but maybe bevy_replicon is trying to send my BIGStruct but it's not making most of the time?

No, this channel should be unreliable because you don't need to resend old data.

**Comment by ActuallyHappening** [comment]
Oh I was intentionally sending a random number, exactly the same as I did from the old bugged version. Using a constant number, I find that the `echo` example never sends it (at least not the twenty times I tried on my local machine)
![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/9ee4da05-1f04-4586-9157-7b892151f727)

This is very interesting, and probably the cause of the issue or at least a 'bug' in of itself.

**Comment by ActuallyHappening** [comment]
This discovery, that larger messages don't send concisely down an Unreliable renet channel, I see has two solutions:
- bevy_replicon adds an API to use reliable channels
- `renet` tries to send larger payloads
- (Or, I redesign my game to use components with less than 40 kilobytes per payload, but I don't want to be restricted by this for no good reason)

**Comment by Shatur** [comment]
> bevy_replicon adds an API to use reliable channels

No, you should never use reliable channel for replication. Replicon tracks what client receives and resends it manually. The problem in Renet.

> renet tries to send larger payloads

This is the root of the problem, you should report this issue to Renet.

> (Or, I redesign my game to use components with less than 40 kilobytes per payload, but I don't want to be restricted by this for no good reason)

You will need to do this too. Otherwise even on small change you will send huge amount of data.

**Comment by ActuallyHappening** [comment]
OK thanks

> > (Or, I redesign my game to use components with less than 40 kilobytes per payload, but I don't want to be restricted by this for no good reason)
> 
> You will need to do this too. Otherwise even on small change you will send huge amount of data.

Not really, my use case is syncing the initial state of the world which can definitely take a lot of data.

**Comment by ActuallyHappening** [comment]
Hey @Shatur, to save creating a new issue, what could the cause of this be?
![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/49a7c9c7-b145-4bef-bf25-84158588211f)

Client:
![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/f5b472fa-ad77-4286-aec8-c0b56bbe34d9)
```
thread 'Compute Task Pool (1)' panicked at /Users/ah/.cargo/registry/src/index.crates.io-6f17d22bba15001f/renet-0.0.13/src/remote_connection.rs:268:13:
Tried to send message to invalid channel 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Encountered a panic in system `bevy_replicon::network_event::client_event::sending_system<creativity_game::bevy::player::thrust::info_gathering::PlayerInputs>`!
Encountered a panic in system `bevy_app::main_schedule::Main::run_main`!
```

Server:
![image](https://github.com/lifescapegame/bevy_replicon/assets/105958073/19c1c6a2-c48c-4aca-ad8f-f74983be2153)


Getting error 'Tried to send message to invalid channel 1' (from client on the right) when evidently the channel configurations allow for channel id 1 on both server (left) and client

Also, is there a better way to chat then over this? email? discord?
